### PR TITLE
[#41] [#114] [Android] [UI] As a user, I can answer my survey

### LINKED ISSUE #41: [Android] [UI] As a user, I can answer my survey
# Why

User should be able to answer questions after start a survey.

# Acceptance Criteria

- Show close (X) button and hide the `Back` button
- Show the current progress (.e.g: 2/5)
- Show the question
- Show awswers or input field base on the type of question.
- Show next button `>`.

Image

[Figma](https://www.figma.com/file/GjRPOjDyZ6f4EDL3wKarRK/Challenge---Mobile-App?node-id=34-765&t=1xiBQA7GNv0kEx7j-0)

### LINKED ISSUE #114: [Android] [UI] As a user, I can answer a question with a list of answer in a spinner
# Why

The user should be able to see and answer a question with a list of answer inside a spinner.

# Acceptance Criteria

- Display a list of answers inside a spinner.
- Highlight the selected value.

Image

### PR DESCRIPTION
**Pull request by AVI5HEK**
- Close #41 
- Close #114 

## What happened 👀

- Show close (X) button and hide the Back button
- Show the current progress (.e.g: 2/5)
- Show the question
- Show awswers or input field base on the type of question.
- Show next button >.

## Insight 📝

- When the user clicks `Start Survey`, the data is fetched from the API and then the user is taken to the questionnaire. For this UI PR, I've used some hardcoded dummy data of 5 duplicate items to test the paging behavior.
- Added the `DROPDOWN` questions to test out the behavior.

## Proof Of Work 📹

https://github.com/nimblehq/avishek-phong-kmm-ic/assets/8093908/e7558dd2-f98c-41a3-90d2-a2f9094e4eac

### CODE REVIEW

**Review by suho — APPROVED** [review]

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/common/CloseButton.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+@Composable
+fun CloseButton(
+    onClick: () -> Unit,
+    modifier: Modifier = Modifier
+) {
+    Image(
+        painter = painterResource(id = R.drawable.ic_close),
+        contentDescription = null,
+        modifier = modifier
+            .size(28.dp)
```
Comment (kaungkhantsoe) [inline androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/common/CloseButton.kt]: I think we can make a constant for this `28.dp` for both close and back button. Also are you planning to reuse these close and back button for other screens?

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/Spinner.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+@Composable
+fun Spinner(
+    modifier: Modifier = Modifier,
+    answerUiModels: List<AnswerUiModel>,
+) {
+    val answerTexts = answerUiModels.map { it.text }
+    var answerText by remember { mutableStateOf(answerTexts[0]) }
+
+    ListItemPicker(
```
Comment (kaungkhantsoe) [inline androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/Spinner.kt:22]: I didn't know this exist. That's really neat 👍 But can we reduce the opacity for the unelected ones? They are currently nearly invisible. And also can you please try to reduce the thickness of the separator?

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt [hunk 3]
```diff
@@ -1,32 +1,60 @@
...
 @Composable
 fun SurveyDetailScreen(
     homeViewModel: HomeViewModel,
     surveyId: String,
-    onBackClick: () -> Unit
+    onBackClick: () -> Unit,
 ) {
     val viewState = homeViewModel.viewState.collectAsStateWithLifecycle()
     val surveyUiModel = viewState.value.surveys.find { it.id == surveyId }
+    var surveyQuestionUiModels by remember { mutableStateOf<List<SurveyQuestionUiModel>>(emptyList()) }
```
Comment (kaungkhantsoe) [inline androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt:35]: Please move the data logic to ViewModel 🙏

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt [hunk 4]
```diff
@@ -1,32 +1,60 @@
...
+                        answerUiModels = List(5) {
+                            AnswerUiModel(
+                                id = (it + 1).toString(),
+                                text = "Text ${it + 1}"
+                            )
+                        }
+                    )
+                }
             },
             onBackClick = onBackClick
```
Comment (kaungkhantsoe) [inline androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt:99]: Cannot preview the screen
Screenshot 2023-05-22 at 12 35 09 PM

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyQuestionContent.kt [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+fun SurveyQuestionContent(
+    surveyQuestionUiModels: List<SurveyQuestionUiModel>,
+    onCloseClick: () -> Unit,
+    modifier: Modifier = Modifier,
+) {
+    val pagerState = rememberPagerState()
+    val scope = rememberCoroutineScope()
+
+    Box(modifier = modifier) {
+        Box(Modifier.fillMaxSize()) {
```
Comment (kaungkhantsoe) [inline androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyQuestionContent.kt]: Redundant `Box`. Please move the content into the first `Box`. 🙏 
```suggestion
```

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyQuestionContent.kt [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+                backgroundColor = White,
+                onClick = onNextClick,
+                modifier = Modifier.size(56.dp)
+            ) {
+                Image(
+                    painter = painterResource(id = R.drawable.ic_arrow_right),
+                    contentDescription = null
+                )
+            }
+        }
```
Comment (kaungkhantsoe) [inline androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyQuestionContent.kt]: We can use `align` for alignment. Don't need to wrap inside another view here.
```suggestion
        FloatingActionButton(
            backgroundColor = White,
            onClick = onNextClick,
            modifier = Modifier.size(56.dp).align(Alignment.End)
        ) {
            Image(
                painter = painterResource(id = R.drawable.ic_arrow_right),
                contentDescription = null
            )
        }
```

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt (hunk 3 above)
Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt:35]: I'd like to do it in the integrate task. Is that okay?

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/common/CloseButton.kt (hunk 1 above)
Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/common/CloseButton.kt]: Removed the file and moved the `Image` to the caller. https://github.com/nimblehq/avishek-phong-kmm-ic/pull/120/commits/f686f7addab4416fb526524b008aed5a74ee2412

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/Spinner.kt (hunk 2 above)
Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/Spinner.kt:22]: I used a 3rd-party library for this. Unfortunately, it doesn't allow hanging the divider thickness and the unselected text styles. We might need to modify the library implementation for our needs. 🤔 How about we do this in a different task?

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt (hunk 4 above)
Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt:99]: Fixed in this commit https://github.com/nimblehq/avishek-phong-kmm-ic/pull/120/commits/f686f7addab4416fb526524b008aed5a74ee2412

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyQuestionContent.kt (hunk 5 above)
Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyQuestionContent.kt]: Resolved in https://github.com/nimblehq/avishek-phong-kmm-ic/pull/120/commits/f686f7addab4416fb526524b008aed5a74ee2412

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyQuestionContent.kt (hunk 6 above)
Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyQuestionContent.kt]: For some reason adding an end padding makes the button narrower. However, used `Box` instead for the same purpose in this commit: https://github.com/nimblehq/avishek-phong-kmm-ic/pull/120/commits/a9b678ce651f4a43e25c7334658eea45de84039a

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt (hunk 3 above)
Comment (kaungkhantsoe) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt:35]: @AVI5HEK For UI task, we can add mutable states inside the ViewModel right? That way I can see how the screen is going to fetch the data before the integration task and I can also see how the data is observed. I think we should implement it in UI task. 🙏

Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyDetailScreen.kt:35]: Ideally, we update the VM and add unit tests in the same integrate PR. But I will update as you said. 🙇🏽

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyQuestionContent.kt (hunk 6 above)
Comment (kaungkhantsoe) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyQuestionContent.kt]: @AVI5HEK I think the order of applying `align`, `size` and `padding` matters. Can you please try applying like the following. I tried it and it turns out the same.
```
modifier = Modifier
                    .padding(end = 20.dp)
                    .align(Alignment.End)
                    .size(56.dp)
```

Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/surveydetail/SurveyQuestionContent.kt]: @kaungkhantsoe you're right. Worked like charm. TIL that the order matters. Only the `size` needs to be the last. The order of `padding` and `align` doesn't matter as it seems.

**Review by kaungkhantsoe — APPROVED** [review]

**Pull request merged**